### PR TITLE
onboarding exercise 4, 5 ,6

### PR DESCRIPTION
**Pull request by rose2221**

### CODE REVIEW

File: rpc/handlers.go [hunk 1]
```diff
@@ -99,8 +102,17 @@ type subscription struct {
 	conn   jsonrpc.Conn
 }
 
+// func RegisterHandlers(rpcServer *jsonrpc.Server, h *Handler) {
```
Comment (rianhughes) [inline rpc/handlers.go:105]: Code that isn't needed / commented out should be removed from the PR

File: rpc/handlers.go [hunk 2]
```diff
@@ -186,15 +199,26 @@ func (h *Handler) Methods() ([]jsonrpc.Method, string) { //nolint: funlen
 			Handler: h.BlockHashAndNumber,
 		},
 		{
-			Name:    "starknet_getBlockWithTxHashes",
+			Name:    "BlockwithTxHashes",
```
Comment (rianhughes) [inline rpc/handlers.go:202]: why did you change the name of this rpc method?

File: rpc/handlers.go [hunk 3]
```diff
@@ -480,4 +509,16 @@ func (h *Handler) MethodsV0_6() ([]jsonrpc.Method, string) { //nolint: funlen
 			Handler: h.Unsubscribe,
 		},
 	}, "/v0_6"
+
+}
+func RegisterHandlers(rpcServer *jsonrpc.Server, h *Handler) error {
```
Comment (rianhughes) [inline rpc/handlers.go:514]: The list of methods defined above are registered elsewhere in the code, no need for this func :)

File: jsonrpc/server.go [hunk 4]
```diff
@@ -161,14 +161,14 @@ func (s *Server) WithListener(listener EventListener) *Server {
...
-		if err := s.registerMethod(methods[idx]); err != nil {
+		if err := s.RegisterMethod(methods[idx]); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func (s *Server) registerMethod(method Method) error {
+func (s *Server) RegisterMethod(method Method) error {
```
Comment (rianhughes) [inline jsonrpc/server.go:171]: why export this function?

File: rpc/block.go [hunk 5]
```diff
@@ -145,6 +148,18 @@ type BlockWithReceipts struct {
 //
 // It follows the specification defined here:
 // https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L500
+
+type BlockWithTxHashesandReceipts struct {
+	Status BlockStatus `json:"status,omitempty"`
+	BlockHeader
+	TxnHashes []*felt.Felt `json:"transactions"`
+	Transactions []TransactionWithReceipt `json:"transactions"`
```
Comment (rianhughes) [inline rpc/block.go:156]: Note: you have two fields with the same json-tag which will cause marshalling errors

File: rpc/block.go [hunk 6]
```diff
@@ -203,12 +218,47 @@ func (h *Handler) BlockWithTxHashesV0_6(id BlockID) (*BlockWithTxHashes, *jsonrp
 	resp.L1DataGasPrice = nil
 	return resp, nil
 }
+func (h *Handler) BlockWithTxHashesandReceipts(id BlockID) (*BlockWithTxHashesandReceipts, *jsonrpc.Error) {
```
Comment (rianhughes) [inline rpc/block.go:221]: Looks good

File: rpc/block_test.go [hunk 7]
```diff
@@ -227,6 +231,173 @@ func TestBlockTransactionCount(t *testing.T) {
 	})
 }
 
+type MockTrie struct {
```
Comment (rianhughes) [inline rpc/block_test.go:234]: Just fyi, there's no need to write mock types by hand. They can be auto-generated from interface definitions. For example, try autogenerating the StateHistoryReader interfaces mock types by executing

go generate -run "mockgen -destination=../mocks/mock_state.go -package=mocks github.com/NethermindEth/juno/core"

(or just clicking "go run generate" above the interface definition if you use VSCode).

This should regenerate mocks/mock_state.go

**Comment by rianhughes** [comment]
Also, when opening a PR for review you should run
1. make format
2. make lint
3. make test
And make sure the tests and lint checks pass

File: rpc/block_test.go [hunk 8]
```diff
@@ -673,14 +844,131 @@ func TestBlockWithReceipts(t *testing.T) {
...
+		assert.Equal(t, latestBlock.Hash, b.Hash)
+		assert.Equal(t, latestBlock.GlobalStateRoot, b.NewRoot)
+		assert.Equal(t, latestBlock.ParentHash, b.ParentHash)
+		assert.Equal(t, latestBlock.SequencerAddress, b.SequencerAddress)
+		assert.Equal(t, latestBlock.Timestamp, b.Timestamp)
+		assert.Equal(t, len(latestBlock.Transactions), len(b.TxnHashes))
+		for i := 0; i < len(latestBlock.Transactions); i++ {
+			assert.Equal(t, latestBlock.Transactions[i].Hash(), b.TxnHashes[i])
+		}
+	}
```
Comment (rianhughes) [inline rpc/block_test.go:894]: You should also check b.Transactions.

File: cmd/juno/juno.go [hunk 9]
```diff
@@ -168,8 +190,91 @@ const (
...
+	v0_6Response bool
+}
+type TracedBlockTransaction struct {
+	TraceRoot       *vm.TransactionTrace `json:"trace_root,omitempty"`
+	TransactionHash *felt.Felt           `json:"transaction_hash,omitempty"`
+}
+const (
+	traceCacheSize     = 128
+)
+type Handler struct {
```
Comment (rianhughes) [inline cmd/juno/juno.go:212]: What is all this code for?

File: rpc/block.go [hunk 10]
```diff
@@ -217,6 +267,19 @@ func (h *Handler) BlockTransactionCount(id BlockID) (uint64, *jsonrpc.Error) {
...
+    convertedKey := h.trie.FeltToKey(key) // Call the feltToKey function
+    nodes, err := h.trie.NodesFromRoot(&convertedKey) // Pass a pointer to NodesFromRoot
+    if err != nil {
+        return nil, &jsonrpc.Error{
+            Code:    jsonrpc.InternalError,
+            Message: err.Error(),
+        }
+    }
+	 return nodes, nil
+}
```
Comment (rianhughes) [inline rpc/block.go:281]: Exercise 5 requires granting the handler access to the global state trie. To do this, try access the headState() using the handlers bcReader. You can then add a method to this interface (eg GlobalStateTrie) to get access to the state trie. Then call GetNodesFromRoot(). You don't need to add a trie field to the handler type, so it can be removed.

File: sync/sync.go [hunk 11]
```diff
@@ -235,36 +239,129 @@ func (s *Synchronizer) verifierTask(ctx context.Context, block *core.Block, stat
...
+				go func() {
+					defer func() { <-latestSem }()
+					highestBlock, err := s.starknetData.BlockLatest(syncCtx)
+					if err != nil {
+						s.log.Warnw("Failed fetching latest block", "err", err)
+					} else {
+						s.highestBlockHeader.Store(highestBlock.Header)
+					}
+				}()
+			default:
```
Comment (rianhughes) [inline sync/sync.go:319]: Exercise 6 requires you to build a new, very simple, sync service. This requires you to retrieve data from the starknet sequencer, namely the new block, state update and any classes that have been declared that block. Once you have this data you should validate it. Once it's validated you can finally Store it in the database. And proceed to do the same for the next block etc.

**Comment by kirugan** [comment]
Discussed with @rose2221 this PR. Closing it.
